Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -10,13 +10,109 @@ extension NSManagedObject {
     }
 }
 
+extension NSEntityDescription {
+    class func insertNewObject<T: NSManagedObject>(
+        into context: NSManagedObjectContext,
+        for request: NSFetchRequest<T>
+    ) -> T {
+        let entity = insertNewObject(forEntityName: T.entityName, into: context)
+        request.entity = NSEntityDescription.entity(forEntityName: T.entityName, in: context)!
+        FetchCache.shared.set(request, objectIds: [entity.objectID])
+        return entity as! T
+    }
+}
+
+protocol NSFetchRequestGettable {}
+
+private let idKey = "id"
+
+extension NSFetchRequestGettable where Self: NSManagedObject {
+    static func fetchRequest(id: String) -> NSFetchRequest<Self> {
+        fetchRequest(keyPath: idKey, equalTo: id)
+    }
+    
+    static func fetchRequest(keyPath: String, equalTo value: String) -> NSFetchRequest<Self> {
+        let request = NSFetchRequest<Self>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return request
+    }
+}
+
+extension NSManagedObject: NSFetchRequestGettable {}
+
 extension NSManagedObject {
+    static func load<T: NSManagedObject>(by id: String, context: NSManagedObjectContext) -> [T] {
+        load(keyPath: idKey, equalTo: id, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(keyPath: String, equalTo value: String, context: NSManagedObjectContext) -> [T] {
+        let request = NSFetchRequest<T>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return load(by: request, context: context)
+    }
+    
     static func load<T>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+        request.entity = NSEntityDescription.entity(forEntityName: entityName, in: context)!
```

Comment:
This looks like it could be a bug waiting to happen; there is no guarantee that `T` will be the same as `Self` but `Self.entityName` used to configure the request. For example, I think I could, in theory, do something like this which IMO is undesirable:

```
let request: NSFetchRequest<ChannelDTO> = NSFetchRequest(entityName: ChannelDTO.entityName)
MessageDTO.load(request, context: someContext)
```

It's probably better to use `Self` here instead of `T` so that the fetch request that's used must return the managed object subclass type that `load` was called on.

Reply:
or, as you've said above:
```
request.entity = T(context: context)
```
would work?

Reply:
Not quite, that line would insert a new object of type `T` into the context.

Reply:
fixed!